src/AlgebraicGeometry/Schemes/CoveredSchemes/Morphisms/Methods.jl

```diff
@@ -35,24 +35,58 @@ function compose(f::AbsCoveredSchemeMorphism, g::AbsCoveredSchemeMorphism)
   Z = codomain(g)
   cf = covering_morphism(f)
   cg = covering_morphism(g)
+  # The general problem here is that the `CoveringMorphism`s for `f` and 
+  # `g` need not have compatible domains and codomains. If they do, we're 
+  # lucky, but if they don't we first need to create the necessary 
+  # refinements and then compose the induced maps.
   if codomain(cf) === domain(cg) 
-    mor_dict = IdDict{AbsSpec, AbsSpecMor}() # TODO: Keep the type of the morphisms?
-    for U in basic_patches(domain(cf))
-      mor_dict[U] = compose(cf[U], cg[codomain(cf[U])])
-    end
-    cc = CoveringMorphism(domain(cf), codomain(cg), mor_dict, check=false)
-    return CoveredSchemeMorphism(X, Z, cc)
+    # The easy case
+    cfg = compose(cf, cg)
+    return CoveredSchemeMorphism(X, Z, cfg)
+  elseif is_refinement(codomain(cf), domain(cg))[1]
+    # Another rather easy case: We only have to put the refinement morphism 
+    # in the middle.
+    ref = refinement_morphism(codomain(cf), domain(cg))
+    cf_ref_cg = compose(cf, compose(ref, cg))
+    return CoveredSchemeMorphism(X, Z, cf_ref_cg, check=true)
+  elseif is_refinement(domain(cg), codomain(cf))[1]
+    # A more tricky case: 
+    #
+    # AxB' --------> B' ---cg---> C
+    #  |             |
+    #  V             V ref
+    #  A --cf------> B
+    #
+    # Here we first have to complete the square to get another 
+    # `CoveringMorphism` representing `f` with `B'` as is codomain.
+    ref = refinement_morphism(domain(cg), codomain(cf))
+    A = domain(cf)
+    B = codomain(cf)
+    C = domain(ref)
+    AxC, to_A, to_C = fiber_product(cf, ref)
+    return CoveredSchemeMorphism(X, Z, compose(to_C, cg), check=true)
   else
-    haskey(refinements(Y), (codomain(cf), domain(cg))) || error("composition of this complicated case is not yet implemented")
-    ref_mor = Y[codomain(cf), domain(cg)] # the refinement `CoveringMorphism`
-    mor_dict = IdDict{AbsSpec, AbsSpecMor}() # TODO: Keep the type of the morphisms?
-    for U in basic_patches(domain(cf))
-      V = codomain(cf[U])
-      W = codomain(ref_mor[V])
-      mor_dict[U] = compose(compose(cf[U], ref_mor[V]), cg[W])
-    end
-    cc = CoveringMorphism(domain(cf), codomain(cg), mor_dict, check=false)
-    return CoveredSchemeMorphism(X, Z, cc)
+    # The most complicated case:
```

Comment:
this is not hit by a test. Can you cook one up?

Reply:
I think, I achieved this. Let's see what test coverage says now.